torch/_dynamo/variables/dicts.py

```diff
@@ -1,45 +1,167 @@
+                Hashable = ConstDictVariable._HashableTracker
+                x = tuple(Hashable(e).underlying_value for e in self.vt.items)
+            else:
+                x = self.vt.as_python_constant()
+            return x
+
+        def __hash__(self):
+            return hash(self.underlying_value)
+
+        @staticmethod
+        def _eq_impl(a: VariableTracker, b: VariableTracker):
+            # TODO: Put this in utils and share it between variables/builtin.py and here
+            if type(a) != type(b):
+                return False
+            elif isinstance(a, tuple):
+                return len(a) == len(b) and all(_eq_impl(u, v) for u, v in zip(a, b))
+            elif is_fake(a):
+                return a is b
+            else:
+                return a == b
+
+        def __eq__(self, other: object) -> bool:
+            Hashable = ConstDictVariable._HashableTracker
+            if not isinstance(other, Hashable):
+                return False
+            return Hashable._eq_impl(self.underlying_value, other.underlying_value)
+
+    def __init__(
+        self, items: Dict[VariableTracker, VariableTracker], user_cls=dict, **kwargs
+    ):
         super().__init__(**kwargs)
 
-        # All the keys are constants
-        assert not any(isinstance(x, VariableTracker) for x in items)
-        self.guards.update(VariableTracker.propagate(items.values())["guards"])
-        self.items = items
+        Hashable = ConstDictVariable._HashableTracker
+
+        # Keys will just be HashableTrackers when cloning, in any other case they'll be VariableTrackers
+        def make_hashable(key):
+            return key if isinstance(key, Hashable) else Hashable(key)
+
+        assert all(
+            isinstance(x, (VariableTracker, Hashable))
+            and isinstance(v, VariableTracker)
+            for x, v in items.items()
+        )
+        self.items = {make_hashable(x): v for x, v in items.items()}
+        self.guards.update(
+            VariableTracker.propagate(
+                [x.vt for x in self.items.keys()], self.items.values()
+            )["guards"]
+        )
         self.user_cls = user_cls
 
     def as_proxy(self):
-        return {k: v.as_proxy() for k, v in self.items.items()}
+        return {k.vt.as_proxy(): v.as_proxy() for k, v in self.items.items()}
 
     def as_python_constant(self):
-        return {k: v.as_python_constant() for k, v in self.items.items()}
+        return {
+            k.vt.as_python_constant(): v.as_python_constant()
+            for k, v in self.items.items()
+        }
+
+    def keys_as_python_constant(self):
+        return {k.vt.as_python_constant(): v for k, v in self.items.items()}
 
     def python_type(self):
         return self.user_cls
 
+    def __contains__(self, obj):
+        # If it's literal we wrap it for you
+        if ConstantVariable.is_literal(obj):
+            obj = ConstantVariable.create(obj)
+
+        if isinstance(obj, VariableTracker):
```

Comment:
Mixing VariableTrackers and non-VariableTrackers in a a single API is confusing an error prone.